Add flag to open docs:  x.py doc --open

dtolnay opened the pull request:
This aligns with Cargo's flag `cargo doc --open`.

Tested with:

```bash
   # opens doc/index.html
x.py doc --stage 0 --open
x.py doc --stage 0 --open src/doc

   # opens doc/book/index.html
x.py doc --stage 0 --open src/doc/book

   # opens doc/std/index.html
x.py doc --stage 0 --open src/libstd

   # opens doc/proc_macro/index.html
x.py doc --stage 0 --open src/libproc_macro

   # opens both
x.py doc --stage 0 --open src/libstd src/libproc_macro
```

dtolnay committed:
Add flag to open docs:  x.py doc --open

Tested with:

       # opens doc/index.html
    x.py doc --stage 0 --open
    x.py doc --stage 0 --open src/doc

       # opens doc/book/index.html
    x.py doc --stage 0 --open src/doc/book

       # opens doc/std/index.html
    x.py doc --stage 0 --open src/libstd

       # opens doc/proc_macro/index.html
    x.py doc --stage 0 --open src/libproc_macro

       # opens both
    x.py doc --stage 0 --open src/libstd src/libproc_macro

rust-highfive commented:
r? @Mark-Simulacrum

(rust_highfive has picked a reviewer for you, use r? to override)

jyn514 reviewed (approved):
Just this week I tried to use this and was surprised it didn't exist :)

Mark-Simulacrum commented on src/bootstrap/doc.rs:
```diff
@@ -70,6 +70,35 @@ book!(
+    }
+
+    let path = path.as_ref();
+    builder.info(&format!("Opening doc {}", path.to_string_lossy()));
```
nit: `.display()`?

dtolnay replied:
Fixed.

Mark-Simulacrum commented on src/bootstrap/doc.rs:
```diff
@@ -70,6 +70,35 @@ book!(
+    builder.info(&format!("Opening doc {}", path.to_string_lossy()));
+
+    // ignore error
+    let _ = opener::open(path);
```
I would have expected us to print the error here - at least on higher levels of verbosity, is there a reason we're ignoring it? I imagine it'd be useful for people who try to find out why open doesn't work for them.

dtolnay replied:
Done. On my machine the error looks like:

```console
    Finished release [optimized] target(s) in 0.29s
Opening doc /git/rust/build/x86_64-unknown-linux-gnu/doc/std/index.html
command 'xdg-open (internal)' did not execute successfully; exit code: 4
command stderr:
gio: file:///git/rust/build/x86_64-unknown-linux-gnu/doc/std/index.html: Error when getting information for file “/git/rust/build/x86_64-unknown-linux-gnu/doc/std/index.html”: No such file or directory

Build completed successfully in 0:00:08
```

dtolnay committed:
Report error from opener in bootstrap

On my machine, an error looks like:

    Finished release [optimized] target(s) in 0.29s
    Opening doc /git/rust/build/x86_64-unknown-linux-gnu/doc/std/index.html
    command 'xdg-open (internal)' did not execute successfully; exit code: 4
    command stderr:
    gio: file:///git/rust/build/x86_64-unknown-linux-gnu/doc/std/index.html: Error when getting information for file “/git/rust/build/x86_64-unknown-linux-gnu/doc/std/index.html”: No such file or directory

    Build completed successfully in 0:00:08

Mark-Simulacrum commented:
@bors r+

Thanks!
